packages/text-bitmap/src/BitmapFontFactory.ts

```diff
@@ -0,0 +1,221 @@
+        {
+            chars = chars.split('');
+        }
+
+        const resolution = options.resolution || 1;
+        const style = new TextStyle(options);
+
+        const lineWidth = BMT_SIZE;
+
+        const fontData = new BitmapFontData();
+
+        fontData.info[0] = {
+            face: style.fontFamily,
+            size: style.fontSize,
+        };
+        fontData.common[0] = {
+            lineHeight: style.fontSize,
+        };
+        fontData.page[0] = {
+            id: 0,
+            file: '',
+        };
+
+        const textures: Texture[] = [];
+
+        let positionX = 0;
+        let positionY = 0;
+
+        let canvas;
+        let context;
+        let baseTexture;
+        let maxCharHeight = 0;
+
+        for (let i = 0; i < chars.length; i++)
+        {
+            if (positionY >= BMT_SIZE - style.fontSize)
+            {
+                canvas = null;
+                context = null;
+                baseTexture = null;
+                positionY = 0;
+                positionX = 0;
+            }
+            if (!canvas)
+            {
+                canvas = document.createElement('canvas');
+                canvas.width = BMT_SIZE;
+                canvas.height = BMT_SIZE;
+
+                context = canvas.getContext('2d');
+                baseTexture = new BaseTexture(canvas, { resolution });
+            }
+
+            const metrics = TextMetrics.measureText(chars[i], style, false, canvas);
+            const width = metrics.width;
+            const height = metrics.height;
+
+            maxCharHeight = Math.max(height + metrics.fontProperties.descent, maxCharHeight);
+
+            if ((width * resolution) + positionX >= lineWidth)
+            {
+                --i;
+                positionY += maxCharHeight * resolution;
+                positionX = 0;
+                maxCharHeight = 0;
+                continue;
+            }
+
+            BitmapFontFactory.render(
+                canvas,
+                context,
+                metrics,
+                positionX,
+                positionY,
+                resolution,
+                style
+            );
+
+            const id = metrics.text.charCodeAt(0);
+
+            textures[id] = new Texture(baseTexture, new Rectangle(positionX / resolution, positionY / resolution,
+                width,
+                height));
+
+            fontData.char[id] = {
+                id,
+                page: id,
+                x: 0,
+                y: 0,
+                width,
+                height,
+                xoffset: 0,
+                yoffset: 0,
+                xadvance: width - (style.dropShadow ? style.dropShadowDistance : 0),
+            };
+            fontData.page[id] = {
+                id,
+                file: '',
+            };
+
+            positionX += (width + (2 * BMT_PADDING)) * resolution;
+        }
+
+        return new BitmapFont(fontData, textures);
+    }
+
+    private static render(
```

Comment:
Even if this is internal, could you add comments to all functions? It's helpful to know what it does without reading the code.

Reply:
Also, how much of this overlaps from Text's rendering?

Reply:
A lot of `drawGlyph` comes from `Text#drawLetterSpacing` and `updateText`. However, it doesn't have the two-pass shadow rendering (to bypass a bug in WebKit).

---
I changed `render` to a local function named `drawGlyph`.